<a href="https://colab.research.google.com/github/larialbu/Calculadora-em-js/blob/main/atividade_aula_07.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


##Obtendo os dados: 
Ao rodarmos a célula abaixo vamos obter os dados diretamente do site:

http://dados.recife.pe.gov.br/tl/dataset/relacao-de-pessoas-vacinadas-covid-19/resource/966e9c4c-df45-40d7-9c58-2f13c61a6d28 


In [9]:
data = pd.read_csv('http://dados.recife.pe.gov.br/dataset/2b5ee58d-9535-4d80-ab32-f63fd475346e/resource/93f4e8c2-430f-4142-86b3-96dab4905b4e/download/casosgravescovid.csv', sep=";")

Utilize os comandos aprendidos em aula para os dados carregados, começando, preferencialmente, por:
- head()
- head(1-2)
- tail()
- info()
- describe()
- seguindo com outros aprendidos para manipulação de tratamento dos dados.

#Limpeza e tratamento dos dados

##Check List de análises

* Analisar a base de dados
* Descartar colunas com poucos registros
* Tratar valores ausentes
* Tratar colunas erradas

#Analise da base de dados

* Observando a estutura da base e suas colunas

In [10]:
display(data.head(1-2))
data.shape

data_notificacao       sexo      raca etnia idade municipio_residencia  \
0           2020-05-30  Masculino    Branca   NaN    53               Recife   
1           2020-06-02   Feminino     Parda   NaN     0               Recife   
2           2020-06-02   Feminino     Parda   NaN    86               Recife   
3           2020-06-01  Masculino    Branca   NaN    66               Recife   
4           2020-06-02  Masculino     Preta   NaN    68               Recife   
...                ...        ...       ...   ...   ...                  ...   
36429       2022-06-08  Masculino    Branca   NaN     1               Recife   
36430       2022-06-08  Masculino  Ignorado   NaN    77               Recife   
36431       2022-06-09  Masculino     Parda   NaN    74               Recife   
36432       2022-06-10  Masculino     Parda   NaN    74               Recife   
36433       2022-06-12   Feminino   Amarela   NaN    73               Recife   

                     bairro distrito_sanitario data_inicio_sintomas  \
0                SAN MARTIN                  V           2020-05-25   
1      ALTO SANTA TEREZINHA                 II           2020-05-30   
2                ESPINHEIRO                III           2020-06-01   
3                CASA FORTE                III           2020-05-25   
4            ILHA DO RETIRO                 IV           2020-05-30   
...                     ...                ...                  ...   
36429             GUABIRABA                VII           2022-06-05   
36430    BREJO DA GUABIRABA                VII           2022-06-06   
36431       NOVA DESCOBERTA                VII           2022-06-06   
36432              BEBERIBE                 II           2022-06-03   
36433       ENGENHO DO MEIO                 IV           2022-06-11   

                                                sintomas outros_sintomas  \
0      Febre, Tosse, Mialgia, Alteração/perda de olfa...             NaN   
1              Febre, Tosse, Dispneia, Saturação O2 < 95             NaN   
2                            Dispneia, Saturação O2 < 95             NaN   
3                                                    NaN             NaN   
4                     Tosse, Dispneia, Saturação O2 < 95             NaN   
...                                                  ...             ...   
36429                             Febre, Tosse, Dispneia             NaN   
36430                                           Dispneia             NaN   
36431  Tosse, Cansaço/fadiga, Desconforto respiratóri...             NaN   
36432                             Febre, Tosse, Dispneia             NaN   
36433  Dispneia, Desconforto respiratório / aperto to...             NaN   

                           doencas_preexistentes  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   
...                                          ...   
36429                                        NaN   
36430                                        NaN   
36431                                        NaN   
36432                                   Diabetes   
36433  Doenças Cardíacas ou Vasculares, Diabetes   

      outras_doencas_preexistentes  profissional_saude categoria_profissional  \
0                               NaN                NaN                    NaN   
1                               NaN                NaN                    NaN   
2                               NaN                NaN                    NaN   
3                               NaN                NaN                    NaN   
4                               NaN                NaN                    NaN   
...                             ...                ...                    ...   
36429                           NaN                NÃO                    NaN   
36430 

(36435, 18)

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36435 entries, 0 to 36434
Data columns (total 18 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   data_notificacao               36435 non-null  object
 1   sexo                           36435 non-null  object
 2   raca                           30106 non-null  object
 3   etnia                          4 non-null      object
 4   idade                          36435 non-null  object
 5   municipio_residencia           36435 non-null  object
 6   bairro                         36435 non-null  object
 7   distrito_sanitario             36435 non-null  object
 8   data_inicio_sintomas           36435 non-null  object
 9   sintomas                       34477 non-null  object
 10  outros_sintomas                5974 non-null   object
 11  doencas_preexistentes          13698 non-null  object
 12  outras_doencas_preexistentes   7807 non-null   object
 13  p

Colunas com poucos registros:

* etinia (apagar)
* outros_sintomas (apagar)
* doencas_preexistentes (analisar)
* outras_doencas_preexistentes (apagar)
* categoria_profissional (apagar)
* data_obito (apagar)

#Descartando colunas

In [12]:

data_copy = data.copy()

* Apagarei as colunas com poucos registros e que nesse caso não tem tanta relevancia.

In [13]:

data.drop(columns = ['etnia','outros_sintomas',\
                     'categoria_profissional','data_obito'], inplace=True)

# Mesmo copiando e colando o nome da coluna "outras_doencas_preexistentes"
# dava erro dizendo que a coluna não existia, então apaguei usando o indice
data.drop(columns = data.columns[10], inplace=True)

* A coluna "doencas_preexistentes" tem metade dos valores ausentes, então será feita uma analise mais profunda para ver se deve ou não ser descartada também.

In [14]:

# Analisando "doencas_preexistentes"
print('* CONTAGEM DE VALORES *'.center(100,'='))
display(data.value_counts('doencas_preexistentes'))
print('* QUANTIDADE VALORES NULOS *'.center(100,'='))
display(data['doencas_preexistentes'].isnull().sum())

======================================* CONTAGEM DE VALORES *=======================================


doencas_preexistentes
Doenças Cardíacas ou Vasculares                                                 3827
Doenças Cardíacas ou Vasculares, Diabetes                                       2633
Diabetes                                                                        1610
Doenças Respiratórias Crônicas                                                   937
Sobrepeso/Obesidade                                                              545
                                                                                ... 
Doenças Hematológicas, Neoplasias                                                  1
Doenças Hematológicas, Imunossupressão, Neoplasias                                 1
Doenças Hematológicas, Imunossupressão, Asma                                       1
Doenças Cardíacas ou Vasculares, Doenças Neurológicas, Doenças Hematológicas       1
Sobrepeso/Obesidade, Neoplasias                                                    1
Length: 244, dtype: int64

====================================* QUANTIDADE VALORES NULOS *====================================


22737

* Apesar de existir muitas opções que podem preencher os dados vazios, isso não seria viável pelo fato de que mais da metade dos valores da coluna são valores nulos. Preencher todos eles poderia criar uma realidade nos dados que não representa a situação real, então a coluna será descartada.

In [15]:
data.drop(columns = ['doencas_preexistentes'], inplace=True)

In [16]:
display(data.head(3))
display(data.shape)
data.info()


data_notificacao       sexo    raca idade municipio_residencia  \
0       2020-05-30  Masculino  Branca    53               Recife   
1       2020-06-02   Feminino   Parda     0               Recife   
2       2020-06-02   Feminino   Parda    86               Recife   

                 bairro distrito_sanitario data_inicio_sintomas  \
0            SAN MARTIN                  V           2020-05-25   
1  ALTO SANTA TEREZINHA                 II           2020-05-30   
2            ESPINHEIRO                III           2020-06-01   

                                            sintomas profissional_saude  \
0  Febre, Tosse, Mialgia, Alteração/perda de olfa...                NaN   
1          Febre, Tosse, Dispneia, Saturação O2 < 95                NaN   
2                        Dispneia, Saturação O2 < 95                NaN   

  classificacao_final                       evolucao  
0          DESCARTADO          ISOLAMENTO DOMICILIAR  
1        INCONCLUSIVO          ISOLAMENTO DOMICILIAR  
2          DESCARTADO  INTERNADO LEITO DE ISOLAMENTO

(36435, 12)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36435 entries, 0 to 36434
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   data_notificacao      36435 non-null  object
 1   sexo                  36435 non-null  object
 2   raca                  30106 non-null  object
 3   idade                 36435 non-null  object
 4   municipio_residencia  36435 non-null  object
 5   bairro                36435 non-null  object
 6   distrito_sanitario    36435 non-null  object
 7   data_inicio_sintomas  36435 non-null  object
 8   sintomas              34477 non-null  object
 9   profissional_saude    27338 non-null  object
 10  classificacao_final   36435 non-null  object
 11  evolucao              35458 non-null  object
dtypes: object(12)
memory usage: 3.3+ MB


#Tratando valores ausentes nas colunas:

* sintomas, profissional_saude e evolucao tem uma baixa quantidade de valores ausentes que podem se tratado.

In [17]:
def resumo_column(coluna):
  print('* VALORES NULOS *'.center(100,'='))
  display(data[coluna].isnull().sum())

  print('* QUANTIDADE VALORES *'.center(100,'='))
  display(data.value_counts(coluna))
  print('total:      ', data.value_counts(coluna).sum())

  print('* VALOR COMUM *'.center(100,'='))
  display(data[coluna].describe().top)
     

##sintomas

In [18]:

resumo_column('sintomas')

=========================================* VALORES NULOS *==========================================


1958

=======================================* QUANTIDADE VALORES *=======================================


sintomas
Febre, Tosse, Dispneia, Saturação O2 < 95                                                                                                  2535
Febre, Tosse, Dispneia                                                                                                                     2306
Febre, Tosse, Dispneia, Desconforto respiratório / aperto torácico, Saturação O2 < 95                                                      1604
Febre, Tosse                                                                                                                                963
Tosse, Dispneia, Saturação O2 < 95                                                                                                          864
                                                                                                                                           ... 
Febre, Tosse, Dispneia, Desconforto respiratório / aperto torácico, Saturação O2 < 95, Alteração/perda de olfato e/ou paladar, 

total:       34477
==========================================* VALOR COMUM *===========================================


'Febre, Tosse, Dispneia, Saturação O2 < 95'

* Neste caso, irei apenas substituir os valores faltantes pelo valor mais comum.

In [19]:
data['sintomas'].fillna(data['sintomas'].describe().top, inplace=True)

print('valores nulos')
display(data['sintomas'].isnull().sum())

valores nulos


0

##profissional_saude

In [20]:
resumo_column('profissional_saude')

=========================================* VALORES NULOS *==========================================


9097

=======================================* QUANTIDADE VALORES *=======================================


profissional_saude
NÃO    24424
SIM     2914
dtype: int64

total:       27338
==========================================* VALOR COMUM *===========================================


'NÃO'

In [21]:

porcentagem_sim = int(data['profissional_saude'].isnull().sum() * 0.10) # A função so aceita int

data['profissional_saude'].fillna('SIM', limit=porcentagem_sim, inplace=True)
data['profissional_saude'].fillna(data['profissional_saude'].describe().top, inplace=True)
     

* Notei que pouco mais de 10% de todos os valores respondidos eram "SIM" então, na substituição, decidi manter esta proporção e divido os valores, fazendo o valor "SIM" receber 10% dos nulos e o valor mais comum receber todo o restante.

In [22]:
resumo_column('profissional_saude')

=========================================* VALORES NULOS *==========================================


0

=======================================* QUANTIDADE VALORES *=======================================


profissional_saude
NÃO    32612
SIM     3823
dtype: int64

total:       36435
==========================================* VALOR COMUM *===========================================


'NÃO'

##evolução

In [23]:
resumo_column('evolucao')

=========================================* VALORES NULOS *==========================================


977

=======================================* QUANTIDADE VALORES *=======================================


evolucao
RECUPERADO                       18931
ÓBITO                             9344
INTERNADO LEITO DE ISOLAMENTO     4344
ISOLAMENTO DOMICILIAR             1463
INTERNADO UTI                     1376
dtype: int64

total:       35458
==========================================* VALOR COMUM *===========================================


'RECUPERADO'

In [24]:
data.dropna(subset = ['evolucao'], inplace=True)

* Por serem dados mais delicados e por serem poucos nulos, apenas larguei os registros com valores vazios.

In [25]:

resumo_column('evolucao')

=========================================* VALORES NULOS *==========================================


0

=======================================* QUANTIDADE VALORES *=======================================


evolucao
RECUPERADO                       18931
ÓBITO                             9344
INTERNADO LEITO DE ISOLAMENTO     4344
ISOLAMENTO DOMICILIAR             1463
INTERNADO UTI                     1376
dtype: int64

total:       35458
==========================================* VALOR COMUM *===========================================


'RECUPERADO'

##raça

In [26]:
# A função diz quantos porcento um determinado valor equivale do outro
def porcento_do_total(quanto_tem, total):
  return round(((quanto_tem*100)/ total) + 0.5)


def porcentagem_da_raca_no_total(raca):
  '''
  Calcula quantos porcento do total a raça passada corresponde.
  '''
  # Pega a quantidade total de valores preenchidos (100%)
  total = data.value_counts('raca').sum()

  # Quantidade de registros com aquela raça
  quanto_tem = dict(data.value_counts('raca'))[raca] 

  # Calcula o % de 100%
  porcentagem = porcento_do_total(quanto_tem, total)
  return porcentagem


def quantidade_limite_raca(raca):
  '''
  Usa a porcentagem correspondente da raça para definir quantos porcento dos 
  nulos vão receber este valor, para manter a proporcionalidade.
  '''
  porcentagem = porcentagem_da_raca_no_total(raca) 
  quantidade = int(round((data['raca'].isnull().sum() * (porcentagem / 100) + 0.5)))
  return quantidade

In [27]:

resumo_column('raca')

=========================================* VALORES NULOS *==========================================


6329

=======================================* QUANTIDADE VALORES *=======================================


raca
Parda       14548
Ignorado     7829
Branca       5472
Preta        1080
Amarela       182
Indígena       18
dtype: int64

total:       29129
==========================================* VALOR COMUM *===========================================


'Parda'

* Aqui temos uma quantidade aceitável de valores nulos para tratamento de substituição, porem, aqui temos duas possíveis soluções: Ou dividimos os nulos entre as três raças mais comuns de maneira "proporcional"; Ou fazemos todos os valores nulos ficarem como "Ignorados".

* Irei dividor a quantidade de nulos ao meio, metade para a primeira solução e metade para a segunda.

In [28]:

# A quantidade é impar então irei arredondar para cima.
limite_parda = quantidade_limite_raca('Parda')
limite_branca = quantidade_limite_raca('Branca')
limite_preta = quantidade_limite_raca('Preta')

# Preenchendo com as três racas mais comuns
data['raca'].fillna('Parda', limit=limite_parda, inplace=True)
data['raca'].fillna('Branca', limit=limite_branca, inplace=True)
data['raca'].fillna('Preta', limit=limite_preta, inplace=True)

# Preenchendo todos o restante de valores nulos com "Ignorado"
data['raca'].fillna('Ignorado', inplace=True)
     

In [31]:
resumo_column('raca')

=========================================* VALORES NULOS *==========================================


0

=======================================* QUANTIDADE VALORES *=======================================


raca
Parda       17713
Ignorado     9536
Branca       6675
Preta        1334
Amarela       182
Indígena       18
dtype: int64

total:       35458
==========================================* VALOR COMUM *===========================================


'Parda'

#Tratando a colunas:

##idade

In [32]:

data.dtypes

data_notificacao        object
sexo                    object
raca                    object
idade                   object
municipio_residencia    object
bairro                  object
distrito_sanitario      object
data_inicio_sintomas    object
sintomas                object
profissional_saude      object
classificacao_final     object
evolucao                object
dtype: object

In [33]:

# Varificando valores exsitentes na coluna
resumo_column('idade')

=========================================* VALORES NULOS *==========================================


0

=======================================* QUANTIDADE VALORES *=======================================


idade
57     633
58     627
68     618
62     611
59     611
      ... 
IGN      3
109      1
112      1
115      1
110      1
Length: 127, dtype: int64

total:       35458
==========================================* VALOR COMUM *===========================================


'57'

* Existem 127 valores diferentes para a idade e alguns são strings.

* Em uma rápida varificação usando loop, ví que existem valores menores que 1 ano, no casos uma contagem de meses.

* Irei substitiuir esses valores por 0, por não ter concluido 1 ano de vida

In [34]:

# Função para trocar os valores.
def subtstituiMesesPor0(valor):
  if not valor.isnumeric():
    valor = 0
  return valor

In [35]:


# A função "apply()" aplica a função passada a cada elemento na série ou dataset
data['idade'] = data['idade'].apply(subtstituiMesesPor0)
data['idade'] = data['idade'].astype(int)

* Agora que temos dados numéricos, se pode ver dados estatisticos sobre essa coluna.

In [36]:

data['idade'].describe()

count    35458.000000
mean        53.553105
std         23.892970
min          0.000000
25%         39.000000
50%         56.000000
75%         71.000000
max        115.000000
Name: idade, dtype: float64

* Usando a função "loc", irei analisar as idade maiores que 100

In [37]:

print('* IDADES E QUANTIDADE *'.center(100,'='))
display(data.loc[(data['idade'] >= 100)]['idade'].value_counts())
print('* TOTAL *'.center(100,'='))
display(data.loc[(data['idade'] >= 100)]['idade'].value_counts().sum())

======================================* IDADES E QUANTIDADE *=======================================


100    23
101    19
102    11
104     7
103     7
105     5
107     5
106     4
114     4
110     1
115     1
112     1
109     1
Name: idade, dtype: int64

=============================================* TOTAL *==============================================


89

* Apesar de poucos, existem casos com idades muito elevadas, então depois de uma breve analise, vou descartar os registros que tenham mais do que 107 anos de idade, já que acima disso não existem muitos registros e que podem ter sido erro.

In [38]:

data = data.loc[(data['idade'] <= 107)]
data['idade'].describe()
     

count    35450.000000
mean        53.539746
std         23.879089
min          0.000000
25%         39.000000
50%         56.000000
75%         71.000000
max        107.000000
Name: idade, dtype: float64

#Resultado.

* No fim, esta é nossa base de dados limpa

In [39]:

data_clean = data.copy

In [44]:

display(data_clean.head(3))
data_clean.shape

AttributeError: ignored

In [ ]:

data_clean.info()